private/protocol/encode.go

```diff
@@ -0,0 +1,95 @@
+package protocol
+
+import (
+	"io"
+)
+
+// A FieldMarshaler interface is used to marshal struct fields when encoding.
+type FieldMarshaler interface {
+	MarshalFields(FieldEncoder) error
+}
+
+// FieldMarshalerFunc is a helper utility that wrapps a function and allows
+// that function to be called as a FieldMarshaler.
+type FieldMarshalerFunc func(FieldEncoder) error
+
+// MarshalFields will call the underlying function passing in the field encoder
+// with the protocol field encoder.
+func (fn FieldMarshalerFunc) MarshalFields(e FieldEncoder) error {
+	return fn(e)
+}
+
+// ValueMarshaler provides a generic type for all encoding field values to be
+// passed into a encoder's methods with.
+type ValueMarshaler interface {
+	MarshalValue() (string, error)
+	MarshalValueBuf([]byte) ([]byte, error)
+}
+
+// A StreamMarshaler interface is used to marshal a stream when encoding.
+type StreamMarshaler interface {
+	MarshalStream() (io.ReadSeeker, error)
+}
+
+// ListMarshaler is a marshaler for list encoders.
+type ListMarshaler interface {
+	MarshalValues(enc ListEncoder) error
```

Comment:
What do you think about naming this interface method `MarshalListValues` ?  To create separation between Field,List, and Map marshaler interfaces.  MarshalMapValues or similar for `MapMarshaler`?